### DISCUSSION TITLE
ftp changes in 0.4.7 - synchronization of file uploads leads to overload

### DISCUSSION THREAD

**Discussion by blfriedman** [roleoroleo/yi-hack-MStar#442] [General]
I have noticed that the method that files are sent to the ftp server (if configured) are different now than they were in 0.4.3.  Previously, as files were created on motion detection, they were sent to the ftp server after the event.  Now, in version 0.4.7, it seems that the file upload occurs on the hour, when a new directory would be created on the camera to store motion events.

Also, if there is a problem with the file upload from the camera to the ftp server, it does not seem like the camera retries to send the directory of motion events, or if it does retry, it does not keep retrying, as I see old directories left on the camera.

The problem is when you have a number of cameras (perhaps 10), and they are all trying to send a large amount of data (a whole hour's worth of captures) up to the same ftp server at the very same time (i.e. on the first minute of the hour).  Occasionally, some of the attempts to ftp the files fail.  Also, if there is a retry period, and it is not randomly backed off, then all units will attempt to retry again, at the same moment.

The old method of sending the files as they were created was superior in producing a randomness to when cameras uploaded files, versus having them all synchronized to send at the same moment.

This issue seems to be related not only to having the same ftp server, but when the cameras are on the same wifi segment.  I have noticed that cameras on other wifi segments did not experience the problem, leading me to believe it is a wifi overload happening at the moment all the cameras on the segment upload in sync.

Has anyone else noticed this?

**Comment by roleoroleo** [comment]
Your analysis is correct apart from 2 points:
1. Now the script is started by cron every 10 minutes (not every hour).
2. If the upload process fails the script tries to send the file again.

You could try to use the startup.sh script to come back to the old method.
Something like:

```
killall crond
cat /var/spool/cron/crontabs/root | grep -v ftppush.sh > /tmp/cron_root
mv /tmp/cron_root /var/spool/cron/crontabs/root
/usr/sbin/crond -c /var/spool/cron/crontabs/
```

**Reply by blfriedman** [reply]
When I look at the crontab, it appears that it is started only at the beginning of the hour - see below:

```
/tmp/var/spool/cron/crontabs # pwd
/var/spool/cron/crontabs
/tmp/var/spool/cron/crontabs # ls
root
/tmp/var/spool/cron/crontabs # cat root
* * * * * /home/yi-hack/script/thumb.sh cron
0 * * * * sleep 20; /home/yi-hack/script/clean_records.sh 75
0 * * * * sleep 40; /home/yi-hack/script/ftppush.sh cron
```

I assume the sleep of 0s, 20s, or 40s is to allow the three cronjobs not to run over each other.

To get the ftp push every 10 minutes, I thought you would need */10 in place of 0 at the beginning of the ftppush crontab line?  Also, when I look at how the cron parameter to ftppush.sh is handled, it seems like it is getting only one shot at sending the file, rather than retrying for the defined sleep interval.

If I understand your suggestion, it appears to  remove ftppush.sh from the root crontab.  How would you suggest to run ftppush - would we simply put "ftppush.sh start" in the /home/yi-hack/script/system.sh startup file?

I appreciate your work and your help with this.

**Reply by roleoroleo** [reply]
> When I look at the crontab, it appears that it is started only at the beginning of the hour - see below:
> I assume the sleep of 0s, 20s, or 40s is to allow the three cronjobs not to run over each other.

I changed it but not commited. The right version is every minute:
https://github.com/roleoroleo/yi-hack-Allwinner-v2/blob/master/src/static/static/yi-hack/script/system.sh

> To get the ftp push every 10 minutes, I thought you would need */10 in place of 0 at the beginning of the ftppush crontab line? Also, when I look at how the cron parameter to ftppush.sh is handled, it seems like it is getting only one shot at sending the file, rather than retrying for the defined sleep interval.

It should retry at the next run.
 
> If I understand your suggestion, it appears to remove ftppush.sh from the root crontab. How would you suggest to run ftppush - would we simply put "ftppush.sh start" in the /home/yi-hack/script/system.sh startup file?

Yes:
```
killall crond
cat /var/spool/cron/crontabs/root | grep -v ftppush.sh > /tmp/cron_root
mv /tmp/cron_root /var/spool/cron/crontabs/root
/usr/sbin/crond -c /var/spool/cron/crontabs/

/home/yi-hack/script/ftppush.sh start &
```

**Comment by blfriedman** [comment]
Thank you.  I have implemented the change to use "ftppush.sh start &" in the system.sh startup script as you suggested - after about a day, it seems to be working great.  I don't see any recordings left on the camera, and all appear to upload correctly.  Looking at /tmp/ftppush.log, it looks to be running properly.  As a failsafe, I do have the cameras reboot daily - so if for any reason the daemon fails, it should get restarted when the camera is rebooted.
